blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +267,225 @@ where
                 .collect::<Result<Vec<Block>, _>>()?;
             // convert tipset to fulltipset
             let full_tipset = FullTipset::new(blocks)?;
-            self.apply_blocks(&full_tipset, &chain_rand)
+            self.apply_blocks(&full_tipset, &chain_rand, callback)
         })
     }
 
+    fn search_back_for_message(
+        &self,
+        tipset: &Tipset,
+        message: &dyn Message,
+    ) -> Result<Option<(Tipset, MessageReceipt)>, Error> {
+        let current = tipset.to_owned();
+
+        if current.epoch() == 0 {
+            return Ok(None);
+        }
+
+        if let Some(actor_state) = self.get_actor(message.from(), tipset.parent_state())? {
```

Comment:
```suggestion
        current: &Tipset,
        message: &dyn Message,
    ) -> Result<Option<(Tipset, MessageReceipt)>, Error> {
        if current.epoch() == 0 {
            return Ok(None);
        }

        if let Some(actor_state) = self.get_actor(message.from(), current.parent_state())? {
```
Swap the names if you want, but the clone should be avoided as tipsets are not trivial to clone